### PR TITLE
Introduce Bifunctor Laws

### PR DESCRIPTION
**Pull request by agenovese**
- Create BifuncorLaws
- Create BifunctorTests
- Add Bifunctor instances for Validated, Ior, Xor, and XorT
- Add Bifunctor tests for Validated, Ior, Xor, and XorT

Fixes: 557

### CODE REVIEW

**Comment by non** [comment]
Thanks! This looks good to me.

:+1:

File: core/src/main/scala/cats/data/Ior.scala [hunk 1]
```diff
@@ -142,6 +144,11 @@ sealed abstract class IorInstances extends IorInstances0 {
     def pure[B](b: B): A Ior B = Ior.right(b)
     def flatMap[B, C](fa: A Ior B)(f: B => A Ior C): A Ior C = fa.flatMap(f)
   }
+
+  implicit def bifunctor: Bifunctor[Ior] =
```
Comment (ceedubs) [inline core/src/main/scala/cats/data/Ior.scala]: Sorry, this is super nitpicky, but could you name this `iorBifunctor` just to match the convention of the instances above it?

Comment (ceedubs) [inline core/src/main/scala/cats/data/Ior.scala]: Similarly for the instances in the other files.

**Comment by ceedubs** [comment]
Other than my super nitpicky comment, this looks good. :+1:

File: core/src/main/scala/cats/data/Ior.scala [hunk 2]
```diff
@@ -142,6 +144,11 @@ sealed abstract class IorInstances extends IorInstances0 {
...
+    new Bifunctor[Ior] {
+      override def bimap[A, B, C, D](fab: A Ior B)(f: (A) => C, g: (B) => D): C Ior D = fab.bimap(f, g)
```
Comment (fthomas) [inline core/src/main/scala/cats/data/Ior.scala]: Could you also remove the parens from `f` and `g`? If functions have only one parameter we usually omit them.

**Comment by fthomas** [comment]
Also LGTM.

**Comment by agenovese** [comment]
Fixed, and sorry about those. I should have noticed the naming and I never use parens like that (going to blame the IDE and the fact that it was after 2am).

**Comment by fthomas** [comment]
Merging, thanks!

**Comment by ceedubs** [comment]
@agenovese no worries. What you had was completely fine! I'm just nitpicky :P

**Pull request merged**